alf/networks/inverse_mvp_network.py

```diff
@@ -0,0 +1,170 @@
+            hidden_size (int): base hidden width for InverseMVP network
+            num_hidden_layers (int): number of hidden layers for InverseMVP net.
+            joint_fc_layer_params: (tuple(int)) widths for each layer in the 
+                InverseMVP network after concatenating the transformed inputs.
+                By default, the width of the joint layers are twice the
+                dimensionality of the input.
+            activation (nn.functional): activation used for hidden layers. The
+                last layer will not be activated.
+            kernel_initializer (Callable): initializer for all the layers but
+                the last layer. If none is provided a variance_scaling_initializer
+                with uniform distribution will be used.
+            name (str):
+        """
+        super().__init__(input_tensor_spec, name=name)
+
+        if isinstance(input_tensor_spec, tuple):
+            assert len(input_tensor_spec) == 2
+            z_spec, eps_spec = input_tensor_spec
+            self._z_dim = z_spec.shape[0]
+            self._eps_dim = eps_spec.shape[0]
+            assert self._eps_dim == output_dim or self._eps_dim == self._z_dim, (
+                "eps_dim must match either z_dim or output_dim!")
+        else:
+            assert isinstance(input_tensor_spec, TensorSpec)
+            self._z_dim = input_tensor_spec.shape[0]
+            self._eps_dim = None
+        assert self._z_dim <= output_dim, (
+            "z_dim cannot be greater than output_dim!")
+
+        self._output_dim = output_dim
+        self._hidden_size = hidden_size
+        self._activation = activation
+
+        if kernel_initializer is None:
+            kernel_initializer = functools.partial(
+                variance_scaling_init,
+                gain=1.0 / 2.0,
+                mode='fan_in',
+                distribution='truncated_normal',
+                nonlinearity=math_ops.identity)
+
+        self._z_encoder = torch.nn.Linear(self._z_dim, hidden_size)
+        if self._eps_dim is not None:
+            self._eps_encoder = torch.nn.Linear(self._eps_dim, hidden_size)
+            joint_hidden_size = 2 * hidden_size
+        else:
+            joint_hidden_size = hidden_size
+        if joint_fc_layer_params is None:
+            joint_fc_layer_params = (joint_hidden_size, ) * num_hidden_layers
+        self._joint_encoder = EncodingNetwork(
+            TensorSpec(shape=(joint_hidden_size, )),
+            fc_layer_params=joint_fc_layer_params,
+            activation=activation,
+            kernel_initializer=kernel_initializer,
+            last_layer_size=output_dim,
+            use_fc_bn=False,
+            last_use_fc_bn=False,
+            last_activation=math_ops.identity)
+
+    def forward(self, inputs, state=()):
+        """Computes prediction given inputs.
+
+        Args:
+            inputs:  Tensor z (self._eps_dim is None) or tuple of Tensors (z, eps) 
```

Comment:
Under what situation is eps not included in `inputs`?

Reply:
Leftover from the minmax-GPVI option that is not included in this PR.
for svgd3, this is never the case. Will change.